add: Dynamic changes to Behaviour Patterns at runtime

SirPigeonz opened the pull request:
This PR makes both FSM and BTree systems work properly when their nodes are moved, removed, added at runtime.

On top of that many nodes are now fine when they don't have important for them nodes as a child or property references. This allows for preparing "empty" slots for intended behaviours or branches and modeling main behaviours with them, that is "slots", in mind.
Upcomming warning system can be used instead for important configuration warnings when behaviours are created by hand in the Editor.

Changes:
- `BTRoot`, `FiniteStateMachine`, `BTComposite`, `BTDecorator`, `FSMStateIntegratedBT` and `BTIntegratedFSM` nodes now can reconfigure themselves when their children are modified at run-time
- allow `BTRoot` to not have any BTBehaviour as a child
- allow `FiniteStateMachine` to not have any FSMStates as a children
- allow `FiniteStateMachine` to not have a initial state
- allow `FSMStateIntegratedBT` to not have any BTRoot as a child
- allow `BTIntegratedFSM` to not have any FSM as a child
- `BTIntegratedFSM` has now default_status property enum for case where it doesn't have FSM child
- new method in `BTRoot` - `swap_entry_point()`
- new method in `FSMStateIntegratedBT` - `swap_behaviour_tree()`
- new method in `BTIntegratedFSM` - `swap_finite_state_machine()`
- make `FSMStates` and extended nodes update their transitions list when children are added/removed
- `FSMStateIntegratedBT` disables autostart of its BTRoot
- `BTIntegratedFSM` disables autostart of its FSM
- made `FSMState` extensible without need to call super() on _ready()
- made `BTRandom` extensible without need to call super() on _ready()
- Added few missing documentation comments, to outline how nodes are intended to work and how they should be edited also at run-time.


Changes after review 1

- removed `_disable_autostart()` in FSM and BTRoot
- made `BTRoot` and `FSM `set their `autostart` to `false` and hide them in Editor Inspector if their parent is a integration type node
- made `BTRoot` and `FSM` a @tool
- added Engine.is_engine_hint guards to ready, callbacks and processes for `BTRoot` and `FSM`
- added `keep_group` optional property to `swap_'nodetype'()` methods, it allows to preserve original nodes groups from swapped node in the new node


Changes after review 2

If you wish to add, remove, move `FSMState` nodes at run-time first add new `FSMStates` stop the FSM with method `FiniteStateMachine.exit_active_state_and_stop` and re-start it with method method `FiniteStateMachine.start` providing one of the new states either as start method property or change member `FiniteStateMachine.initial_state` before running `start()`. After this procedure you can delete unused states.

- made `active` property read-only
- modified `start()` in fsm.gd to accept `FSMState` property as a start point
- new method exit_active_state_and_stop() to pair with `start()`
- above two changes make FSM startable and stoppable for example to safely modify `FSMStates` and resume running of the FSM
- removed some `if` guards from proccess function and made checks when something changes in the setup
- made BTRoot cleanups and made it to not check for entry point in processing
- changed some configuration warnings. Mostly changed statement that state "nodes must have child nodes", to "nodes SHOULD have child nodes to work" to inform user that they wont work but nothing bad will happen if they don't
- removed warning for `BTLeaf` that "BTLeaf node must not have any children.". Reason is that there is no issue if it has one, it can prevent user to use some nice composition on top of `BTLeaf` for no reason :)

ThePat02 commented:
Keep in mind that some things might change how everything is handled, since #17 requires every script to be a `@tool`!

SirPigeonz commented:
@ThePat02
Finished the PR ready for review :)

SirPigeonz commented:
@ThePat02 
Ready for review :) 

This time for sure xD

ThePat02 commented:
Good work! I'll look into it when I finished implementing #17 to avoid any breaking conflicts.

SirPigeonz commented:
> Good work!
Senkyu! :3

> I'll look into it when I finished implementing #17 to avoid any breaking conflicts.
Do as you prefer :)
Although, I would personally do opposite, because I changed some behaviours of the addon. Some configurations that were illegal and crash in the past now are legal just do nothing and would need different type of warning and asserts. (I'm just worried that some of your work will be not needed).

That said I tried to test it diligently as I could. Each node was tested alone and with others and main example also runs as used to.
I also plan to move to this version (2.0 + this PR) with my prototype so I will probably also catch regressions.

What I want the most if you to look into the desing to tell me what you think and if I missed something or should change something :)

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/fsm.gd:
```diff
@@ -40,7 +38,9 @@ signal state_changed(state: FSMState)
 @export var blackboard: Blackboard
 
-
+## Whether the FSM is active or not.[br]
+## To activate use the [code]start()[/code] method.
+var active: bool = false
```
I guess it would make sense to change it to `_active`!

SirPigeonz replied:
I had the same though actually! Then I did it... then I realized some nodes actually need and do call `active`... so it's not only internal property, not sure what to do.

ThePat02 replied:
Then let's just leave it at that for now.

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/fsm.gd:
```diff
@@ -51,16 +51,19 @@ var current_events: Array[String]
-func _ready() -> void:
-	connect("state_changed", _on_state_changed)
 
+# Connecting signal using @onready to omit the need to use super() call
+# in _ready() of extended nodes if they override _ready().
+@onready var __connect_child_order_changed: int = \
+	child_order_changed.connect(_update_states_list)
```
I am not a fan of using `\` to wrap code in such case as I think it makes the code very messy tbh

SirPigeonz replied:
Same, but what other choice we have? If I make it shorter it will still be to long or cryptic. Scrolling way right is also not very nice solution. I made it like this because it's the lesser evil but I'm open for suggestions.

ThePat02 replied:
Are you using the Godot Editor? In VSCode this line is easily fully visible and doesn't need any scrolling.

SirPigeonz replied:
> Are you using the Godot Editor? In VSCode this line is easily fully visible and doesn't need any scrolling.

Ah, no, no, I have a lot of space to see even very long lines past "100 characters line" even xD (most of the time):
![image](https://github.com/ThePat02/BehaviourToolkit/assets/6152134/7b634795-76ee-49a9-84d8-c043e6efc03b)

I just break lines so they don't have more thatn 80 characters like PEP8 demands it xD
It is usefull tho, when you want to work in split file view or make editor snap to half of your monitor.

I can follow different rules I just did what is usually done in python files. Lesser evil although I got used to it too.

squk replied:
Drive-by: I'd recommend enforcing use of this [formatter](https://github.com/Scony/godot-gdscript-toolkit_

SirPigeonz replied:
> Drive-by: I'd recommend enforcing use of this [formatter](https://github.com/Scony/godot-gdscript-toolkit)

Oh! This cool I will use it for my projects for sure, thx for the link!

ThePat02 replied:
> Drive-by: I'd recommend enforcing use of this [formatter](https://github.com/Scony/godot-gdscript-toolkit)

Works like a charm. I'll use it after most 2.0 features are merged. Maybe I'll set up some github actions to do it too!

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/fsm_state.gd:
```diff
@@ -7,10 +7,12 @@ class_name FSMState extends BehaviourToolkit
-	for transition in get_children():
-		if transition is FSMTransition:
-			transitions.append(transition)
+# Connecting signal using @onready to omit the need to use super() call
+# in _ready() of extended nodes if they override _ready().
+@onready var __connect_update_transitions: int = \
+	child_order_changed.connect(_update_transitions)
+@onready var __connect_update_transition_on_ready: int = \
+	ready.connect(_update_transitions)
```
Same as comment above!

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/fsm_state_integrated_bt.gd:
```diff
@@ -1,30 +1,68 @@
+@export var event: String
+
+
+# Connecting signal using @onready to omit the need to use super() call
+# in _ready() of extended nodes if they override _ready().
+@onready var __connect_disable_autostart: int = \
+	ready.connect(_disable_autostart)
+@onready var __connect_child_order_changed: int = \
+	child_order_changed.connect(_on_child_order_changed)
```
Same as comment above.

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/fsm_state_integrated_bt.gd:
```diff
@@ -1,30 +1,68 @@
+		add_child(btree, force_readable_name)
+	else:
+		behaviour_tree.queue_free()
+		add_child(btree, force_readable_name)
```
Just realised you could do this. When adding nodes using the UI, I iterate over the nodes manually. May change this in the future.

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/fsm_state_integrated_bt.gd:
```diff
@@ -37,3 +75,27 @@ func _get_behaviour_tree() -> BTRoot:
 	return null
+
+
+func _disable_autostart() -> void:
+	if Engine.is_editor_hint():
+		return
+	
+	# To prevent wrong configuration from the user side disable autostart.
+	# States BTree should be run only when state is active.
+	# TODO make editor inspector disable autostart when BTree is a child
+	# of FSMStateIntegratedBT?
+	if get_child_count() == 0:
+		return
+	
+	for child in get_children():
+		if child is BTRoot:
+			child.autostart = false
+
```
Is this really necessary? Autostart is set to `false` on default. Maybe using a configuration warning is more appropriate here. 

> make editor inspector disable autostart when BTree is a child

This is also a good idea. The inspector plugin can hide the autostart option and set it to `false` automatically.

ThePat02 reviewed (commented):
The FSM looks good. I'll check the BT later and do some testing!

ThePat02 reviewed (commented):
One other thing I noticed: When changing states of a FSM, what happens if you change the active state? Did the current state exit?

SirPigeonz commented:
@ThePat02 
Adressed few points in new commit:

- removed `_disable_autostart()` in FSM and BTRoot
- made `BTRoot` and `FSM `set their `autostart` to `false` and hide them in Editor Inspector if their parent is a integration type node
- made `BTRoot` and `FSM` a @tool
- added Engine.is_engine_hint guards to ready, callbacks and processes for `BTRoot` and `FSM`
- added `keep_group` optional property to `swap_'nodetype'()` methods, it allows to preserve original nodes groups from swapped node in the new node

ThePat02 commented:
> > One other thing I noticed: When changing states of a FSM, what happens if you change the active state? Did the current state exit?
> 
> Good question! Hmm, I should add active_state guard for processing too.
> 
> When you change state with `change_state()` it runs states `exit` same with transitions. Or do you mean is it running exit when deleted with `Node` API? Then no it doesn't run `exit`. Should it?
> 
> We could add method to `FiniteStateMachine` for deleting states that calls `exit` if state is current active_state, for situations it's needed.
> 
> What do you think?

Well what is the workflow with dynamic changing of nodes? There needs to be some kind of failsafe when changin states manually and via code.

ThePat02 commented:
I think it is best to push this feature to a later version of the plugin (possibliy 2.1) as there are some other pitfalls and problems that come with this. I think it is very important to release the new syntax changes, as more and more people are interested in using it. I'd like to avoid problems like people having to rewrite their code with the new syntax too much.

SirPigeonz commented:
> I think it is best to push this feature to a later version of the plugin (possibliy 2.1) as there are some other pitfalls and problems that come with this. I think it is very important to release the new syntax changes, as more and more people are interested in using it. I'd like to avoid problems like people having to rewrite their code with the new syntax too much.

I'm fine if it will be in 2.1 if you want to release 2.0 early due to changes in API. I use latest code in my game either way.

Furthermore, I just finished a bigger commission for a client, so I will have more free time now to work on this and my game. I will probably finish a new version of this PR today or tomorrow.

ThePat02 commented:
Sounds great. If you need anything, make sure to add me. I'll unsub for now so I don't get notis for every commit!

SirPigeonz commented:
@ThePat02 
Ready for review again :)

ThePat02 commented:
I'll take a look at it after finishing up the other changes. BTW I added a formatting action you can steal for your own projects.
